lib/puppet/parser/functions/pw_hash.rb

```diff
@@ -0,0 +1,50 @@
+Puppet::Parser::Functions::newfunction(
+  :pw_hash,
+  :type => :rvalue,
+  :arity => 3,
+  :doc => "Hashes a password using the crypt function. Provides a hash
+  usable on most POSIX systems.
+
+  The first argument to this function is the password to hash. If it is
+  undef or an empty string, this function returns undef.
+
+  The second argument to this function is a specifier of which hash type
+  to use. Valid hash types are:
+
+  |Specifier|Hash type            |
+  |---------|---------------------|
+  |1        |MD5                  |
+  |5        |SHA-256              |
+  |6        |SHA-512 (recommended)|
+
+  The third argument to this function is the salt to use.
+
+  Note: this uses the Puppet Master's implementation of crypt(3). If your
+  environment contains several different operating systems, ensure that they
+  are compatible before using this function.") do |args|
+    raise ArgumentError, "pw_hash(): wrong number of arguments (#{args.size} for 3)" if args.size != 3
+    raise ArgumentError, "pw_hash(): first argument must be a string" unless args[0].is_a? String
+    raise ArgumentError, "pw_hash(): #{args[1]} is not a valid hash type" unless ['1', '5', '6'].include? args[1].to_s
```

Comment:
Can we do this the other way around?

User-understanding wise I much prefer: `pw_hash('lalalala', 'sha-256', 'oseatnoirsentaiore')` over `pw_hash('lalalala', '5', 'oseatnoirsentaiore')`.

The first case I can immediately see what's going on. The second case I need to look up some implementation detail around specifiers to understand what type of hash this function call will be computing.